presto-hive/src/main/java/io/prestosql/plugin/hive/s3/PrestoS3FileSystem.java

```diff
@@ -795,15 +799,86 @@ public PrestoS3InputStream(AmazonS3 s3, String host, Path path, int maxAttempts,
             closeStream();
         }
 
+        @Override
+        public int read(long position, byte[] buffer, int offset, int length)
+                throws IOException
+        {
+            checkClosed();
+            if (position < 0) {
+                throw new EOFException(FSExceptionMessages.NEGATIVE_SEEK);
+            }
+            checkPositionIndexes(offset, offset + length, buffer.length);
+            if (length == 0) {
+                return 0;
+            }
+
+            try {
+                return retry()
+                        .maxAttempts(maxAttempts)
+                        .exponentialBackoff(BACKOFF_MIN_SLEEP, maxBackoffTime, maxRetryTime, 2.0)
+                        .stopOn(InterruptedException.class, UnrecoverableS3OperationException.class, EOFException.class)
+                        .onRetry(STATS::newGetObjectRetry)
+                        .run("getS3Object", () -> {
+                            InputStream in;
+                            try {
+                                GetObjectRequest request = new GetObjectRequest(host, keyFromPath(path))
+                                        .withRange(position, (position + length) - 1);
+                                in = s3.getObject(request).getObjectContent();
+                            }
+                            catch (RuntimeException e) {
+                                STATS.newGetObjectError();
+                                if (e instanceof AmazonS3Exception) {
+                                    switch (((AmazonS3Exception) e).getStatusCode()) {
+                                        case HTTP_RANGE_NOT_SATISFIABLE:
+                                            throw new EOFException(FSExceptionMessages.CANNOT_SEEK_PAST_EOF);
+                                        case HTTP_FORBIDDEN:
+                                        case HTTP_NOT_FOUND:
+                                        case HTTP_BAD_REQUEST:
+                                            throw new UnrecoverableS3OperationException(path, e);
+                                    }
+                                }
+                                throw e;
+                            }
+
+                            STATS.connectionOpened();
+                            try (Closeable ignored = in) {
+                                int read = 0;
+                                while (read < length) {
+                                    int n = in.read(buffer, offset + read, length - read);
+                                    if (n <= 0) {
+                                        break;
+                                    }
+                                    read += n;
+                                }
+                                return read;
+                            }
+                            catch (IOException e) {
+                                STATS.newReadError(e);
+                                abortStream(in);
```

Comment:
here stream should already be closed. Do we need to additionally abort it?

Reply:
You're right. I was thinking it was closed after the catch block, but it happens before. I'll change to normal try block and do the close at the end.